Enhancement: add redis sentinel support

mike12806 opened the pull request:
<!--
Note: All PRs with code changes should be targeted to the `dev` branch, pure documentation changes can target `main`
-->

## Proposed change

<!--
Please include a summary of the change and which issue is fixed (if any) and any relevant motivation / context. List any dependencies that are required for this change. If appropriate, please include an explanation of how your proposed change can be tested. Screenshots and / or videos can also be helpful if appropriate.
-->

Redis Sentinel offers an easy way to deploy a Redis cluster in high availability, by deploying a new redis-sentinel service (a few instances of it), monitoring the status of different Redis instances set in master/replica mode. If the master goes down, the Sentinel cluster transparently transfers responsibility to a current replica, allowing applications to keep chugging along.

**Sentinel Support:**

redis-py 5.2.1: Native Sentinel support has been in redis-py since v2.10.0 (2016). Version 5.x has mature, stable Sentinel implementation.

Celery 5.5.3: Full Sentinel support via kombu.transport.redis.SentinelChannel - confirmed available.

channels-redis 4.3.0: Uses redis-py's connection classes, so it automatically inherits Sentinel support when you pass a Sentinel-created connection.

**Why It's Safe:**

- Uses the existing redis[hiredis] package already in your dependencies
- No new libraries added
- My implementation uses standard redis-py APIs that have been stable for years
- Full backward compatibility - existing Redis URLs continue to work unchanged
- The Sentinel code paths only activate when Sentinel env vars are set

**I used Github Copilot with Claude Sonnet 4/4.5 to assist in creating this code**

<!--
⚠️ Important: Pull requests that implement a new feature or enhancement *should almost always target an existing feature request* with evidence of community interest and discussion. This is in order to balance the work of implementing and maintaining new features / enhancements. If that is not currently the case, please open a feature request instead of this PR to gather feedback from both users and the project maintainers.
-->

Closes #10526

## Type of change

<!--
What type of change does your PR introduce to Paperless-ngx?
NOTE: Please check only one box!
-->

- [ ] Bug fix: non-breaking change which fixes an issue.
- [X] New feature / Enhancement: non-breaking change which adds functionality. _Please read the important note above._
- [ ] Breaking change: fix or feature that would cause existing functionality to not work as expected.
- [ ] Documentation only.
- [ ] Other. Please explain:

## Checklist:

<!--
NOTE: PRs that do not address the following will not be merged, please do not skip any relevant items.
-->

- [X] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [X] If applicable, I have included testing coverage for new code in this PR, for [backend](https://docs.paperless-ngx.com/development/#testing) and / or [front-end](https://docs.paperless-ngx.com/development/#testing-and-code-style) changes.
- [X] If applicable, I have tested my code for breaking changes & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [X] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [X] I have run all `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [X] I have made corresponding changes to the documentation as needed.
- [X] In the description of the PR above I have disclosed the use of AI tools in the coding of this PR.

mike12806 committed:
feat: add redis sentinel support

mike12806 committed:
chore: pre-commit hooks

mike12806 committed:
fix: resolve CI test failures for Redis Sentinel PR

- Fix system status view to properly import _get_redis_connection helper
- Return actual Redis URL instead of placeholder in status endpoint
- Fix wait-for-redis test to use relative path instead of hardcoded absolute path
- Use patch.dict properly in integration tests to avoid environment pollution
- Remove manual environment cleanup since patch.dict handles it automatically

shamoon commented:
Thats kind of debatable. We tend to think features with votes in the hundreds as popular (of course balanced against the scope of the PR). It does feel a little niche, and it's not like we're getting reports of redis being brought down by paperless. Beyond the complexity of the AI-generated code (manageable or not) there is the issue of needing to be able to test against sentinel, ensuring continued compatibility etc. So I too am a bit worried that basically the 'risks outweigh the benefits'

mike12806 commented:
> Thats kind of debatable. We tend to think features with votes in the hundreds as popular (of course balanced against the scope of the PR). It does feel a little niche, and it's not like we're getting reports of redis being brought down by paperless. Beyond the complexity of the AI-generated code (manageable or not) there is the issue of needing to be able to test against sentinel, ensuring continued compatibility etc. So I too am a bit worried that basically the 'risks outweigh the benefits'

No worries. TBH I won't spend much more time on this if there's this much resistance. Unfortunately it's not about Paperless taking down the redis instance, but the risk of data loss during ingestion. 

I do think the bias against AI assisted coding is a bit odd, but to each their own. Why ask folks to proactively disclose if AI is being used in the PR only to then use it as a reason to push back on their changes?

mike12806 committed:
fix: correct test indentation for malformed_sentinel_hosts test

The test was calling _parse_redis_sentinel_config() outside the patch.dict
context, causing it to run without the required environment variables set.

mike12806 committed:
refactor: reduce code duplication in Redis Sentinel implementation

- Extract duplicate tearDown methods into RedisSentinelTestMixin
- Define _DEFAULT_REDIS_URL constant instead of duplicating literal
- Addresses SonarCloud code quality feedback (6.5% -> <3% duplication)

mike12806 committed:
refactor: eliminate code duplication by extracting shared Redis Sentinel utilities

- Created redis_sentinel_utils.py with shared parse_redis_sentinel_config()
- Updated settings.py, wait-for-redis.py, and test_wait_for_redis.py to use shared utility
- Reduces duplication from 35.7% to near zero in wait-for-redis.py
- Reduces duplication from 15.9% to minimal in settings.py
- Reduces duplication from 9.7% to minimal in test_wait_for_redis.py
- All 32 Redis Sentinel tests passing

mike12806 committed:
fix: address SonarCloud code quality issues

- Extract cleanup logic to _cleanup_sentinel_env_vars() helper method
- Add noqa comments for intentional E402 import order in wait-for-redis.py
- All pre-commit hooks passing

mike12806 committed:
fix: add noqa comment for tearDown method naming in test_redis_sentinel.py

- Suppress N802 warning for framework-required tearDown method name

mike12806 committed:
fix: suppress SonarCloud security warnings for test fixture passwords

- Add nosec comments to test password literals in redis_pass references
- These are non-sensitive test fixtures, not real credentials
- Addresses SonarCloud false positives for hardcoded credential detection

mike12806 committed:
chore: remove nosec comments

- Project doesn't use Bandit, uses Ruff for linting
- nosec comments don't suppress SonarCloud security hotspots
- SonarCloud hotspots require manual review in the dashboard

mike12806 commented:
> The points above are definitely worth considering.
> 
> > I do think the bias against AI assisted coding is a bit odd, but to each their own. Why ask folks to proactively disclose if AI is being used in the PR only to then use it as a reason to push back on their changes?
> 
> We ask because otherwise people don't disclose it, and our goal is definitely not to generically "push-back" against using AI for PRs, nor is it somehow inherently disqualifying or something. But in the last 6–12 months we (as well as every FOSS project out there) have seen a massive influx of PRs from authors who cannot understand, let alone support the code they submit, not to mention the fundamental quality issues some have. We've seen it many times by now. In short, not all contributors use AI carefully. I think all FOSS maintainers are any least concerned about this and where it's headed.

That's fair; definitely appreciate the work you do and I'm sure it must be challenging with the additional volume of PRs enabled by AI. 

FWIW; did spend some additional time here today and looks like tests are passing. I've also pulled the docker image into my K8s env locally and it's working with my existing Sentinel setup. The SonarCloud alerts can be supressed as the entries for passwords are placeholders - I couldn't find a method to supress those in code but I assume they can be supressed by an admin here? 

As @Kernald mentioned, [Sentinel](https://redis.io/docs/latest/operate/oss_and_stack/management/sentinel/) is a pretty widely adopted approach so I'd hope there's low risk of support being dropped from the libs that are ultimately enabling Sentinel here. 

Last point I'll make (and then I'll leave it up to y'all to decide) is that I wouldn't be pushing the addtn'l complexity if Paperless-ngx wasn't so critical to myself (and I imagine others) from the perspective of wanting to avoid data loss. While most things in my homelab are "nice to have", I definitely put Paperless in the "critical" category given the nature of the docs I keep inside Paperless. 

Thanks again for your time!

mike12806 committed:
chore: add sonar-project.properties to suppress test security hotspots

mike12806 committed:
style: fix trailing whitespace in sonar-project.properties

mike12806 committed:
chore: suppress SonarCloud naming convention warning for test methods

mike12806 committed:
chore: broaden SonarCloud exclusions for test files

mike12806 committed:
chore: add inline SonarCloud suppressions for test file

mike12806 committed:
chore: add inline SonarCloud suppressions for test_wait_for_redis.py

mike12806 committed:
Merge remote-tracking branch 'upstream/dev' into feat/add-redis-sentinel-support

mike12806 committed:
Merge remote-tracking branch 'upstream/dev' into feat/add-redis-sentinel-support

mike12806 committed:
Merge branch 'dev' into feat/add-redis-sentinel-support

mike12806 committed:
Merge remote-tracking branch 'upstream/dev' into feat/add-redis-sentinel-support

mike12806 commented:
Hey @shamoon - just wanted to check in on this one - thanks!
